google-cloud-storage/src/main/java/com/google/cloud/storage/BufferToDiskThenUpload.java

```diff
@@ -211,41 +215,78 @@ public ApiFuture<BlobInfo> getResult() {
         return result;
       }
 
+      @SuppressWarnings("UnstableApiUsage")
       private final class Flusher implements WritableByteChannel {
 
         private final WritableByteChannel delegate;
+        private final Hasher cumulativeCrc32c;
+        private final ReentrantLock lock;
 
         private Flusher(WritableByteChannel delegate) {
           this.delegate = delegate;
+          this.cumulativeCrc32c =
+              opts.getHasher().initialValue() == null ? null : Hashing.crc32c().newHasher();
+          this.lock = new ReentrantLock();
         }
 
         @Override
         public int write(ByteBuffer src) throws IOException {
-          return delegate.write(src);
+          lock.lock();
+          try {
+            if (cumulativeCrc32c != null) {
+              cumulativeCrc32c.putBytes(src.duplicate());
+            }
+            return delegate.write(src);
+          } finally {
+            lock.unlock();
+          }
         }
 
         @Override
         public boolean isOpen() {
-          return delegate.isOpen();
+          lock.lock();
+          try {
+            return delegate.isOpen();
+          } finally {
+            lock.unlock();
+          }
         }
 
         @Override
         public void close() throws IOException {
-          delegate.close();
-          try (RecoveryFile rf = Factory.WriteToFileThenUpload.this.rf) {
-            Path path = rf.getPath();
-            long size = Files.size(path);
-            ThroughputSink.computeThroughput(
-                clock,
-                gcs,
-                size,
-                () -> {
-                  BlobInfo blob = storage.internalCreateFrom(path, info, opts);
-                  result.set(blob);
-                });
-          } catch (StorageException | IOException e) {
-            result.setException(e);
-            throw e;
+          lock.lock();
```

Comment:
maybe a dumb question, but why does this need a lock now if it didn't need one before?

Reply:
Previously, we didn't have any state we need to keep in sync with what was being written to disk in `write`. Now that we're hashing before writing to disk I want to make sure that our hashing is made threadsafe.